Particle swarm

bosonie opened the pull request:
Implement the Particles-Swarm optimization engine as described in http://dx.doi.org/10.1109/CEC.2003.1299391.

To test the implementation one can use this code:
```
from aiida.engine import WorkChain
from aiida.engine.launch import run
from aiida.orm import List,Float,Dict

from aiida_optimize import OptimizationWorkChain
from aiida_optimize.engines import ParticleSwarm

class x2y2(WorkChain):
    """
    A simple workchain which represents the function to be optimized.
    """
    @classmethod
    def define(cls, spec):
        super(x2y2, cls).define(spec)

        spec.input('x', valid_type=List)
        spec.output('result', valid_type=Float)

        spec.outline(cls.evaluate)

    def evaluate(self):
        # This is a bit improper: The new value should be created in a calculation.
        self.out('result', Float(self.inputs.x.get_list()[0]**2+self.inputs.x.get_list()[1]**2).store())

result_wc = run(
    OptimizationWorkChain,
    engine=ParticleSwarm,
    engine_kwargs=Dict(dict=dict(particles=[[1.3,1.],[-1.,-2.],[22,3],[-3.,-4.]], max_iter=50, result_key='result')),
    evaluate_process=x2y2
)

print('\nResult with workchain:', result_wc)
```

@greschd the implementation is very basic, and, to be noted, at the moment does not have a check on the function values or the variables distance to decide when to stop. The optimization always goes on until the `max_iter` is reached. This is pretty standard for particle-swarm, but we can instead add to have checks very similar to the simplex engine. What is your opinion on that?
Also, there is a bit of repetition of code among this implementation and the simplex one.

I also pin here @vdikan. Vladimir, I would like you can quickly review the algorithm, just to make sure I did not do silly mistakes. Also, I did not have time to implement the penalty function. I believe you can add it later on, after this PR is merged. It should be very easy to do.

bosonie committed:
First implementation of the Particle-Swarm optimization engine

bosonie committed:
Improve comments and report strings

greschd commented on aiida_optimize/engines/_particle_swarm.py:
```diff
@@ -0,0 +1,299 @@
+# -*- coding: utf-8 -*-
+
+# ******NOTICE***************
```
I don't think this notice is needed here - it's included in the `NelderMead` code because I copied the algorithm. Here, the AiiDA specific things are copied from `NelderMead`.

greschd commented on aiida_optimize/engines/_particle_swarm.py:
```diff
@@ -0,0 +1,299 @@
+    def __init__(  # pylint: disable=too-many-arguments
+        self,
+        particles: ty.List[float], #ty.Optional[ty.List[float]],
+        #xtol: ty.Optional[float],
```
I guess the commented-out code can be removed (here and in some other places)?

bosonie replied:
Yes, it was just the part that introduces a tolerance, in case you wanted it immediately

greschd reviewed (commented):
Thanks for the contribution @bosonie! I think this is mostly good to go -- the interface looks good to me, which is the most important part.

> To test the implementation one can use this code

Would you mind turning this into a simple test? See https://github.com/greschd/aiida-optimize/blob/develop/tests/test_nelder_mead.py for an example, the `x2y2` workchain can go here: https://github.com/greschd/aiida-optimize/blob/develop/tests/sample_processes.py

> at the moment does not have a check on the function values or the variables distance to decide when to stop

That fine with me, we can always add an optional stopping criterion later if needed.

> Also, there is a bit of repetition of code among this implementation and the simplex one.

Yeah, that's also good for me. I wouldn't try to abstract this before we have 3-4 engines using the same approach.

The pre-commit and docs failures seem unrelated to this change, I'll have a look at that.

greschd committed:
Merge branch 'develop' into ParticleSwarm

greschd commented:
> The pre-commit and docs failures seem unrelated to this change, I'll have a look at that.

The continuous integration is fixed, I've taken the liberty to merge it into this branch.

bosonie committed:
Polishing of code and introduction of a test

bosonie commented:
@greschd it should be ready, please note I had to change the `check_optimization` in `conftest.py` in order to make some inputs optional. Have a look if it is good!

bosonie committed:
Reduce again the requirement for particle swarm test to converge

greschd commented on tests/test_particle_swarm.py:
```diff
@@ -0,0 +1,39 @@
+        engine_kwargs=dict(particles=particles, max_iter=15),
+        func_workchain_name=func_workchain_name,
+        ftol=0.001,
+        x_exact=x_exact,
```
AFAICT `x_exact` has no effect if `xtol` is `None` now.

greschd commented on tests/conftest.py:
```diff
@@ -64,10 +64,10 @@ def inner(  # pylint: disable=too-many-arguments,missing-docstring,useless-suppr
-        ftol,
-        x_exact,
-        f_exact,
+        xtol=None,
```
I don't completely understand why these need to be made optional: they are the exact value and tolerance of the expected result, not parameters passed to the optimization engine.
AFAICT, we can have a known result for the optimization even if we don't pass tolerances to the engine itself -- we can make the tolerances "large enough" to ensure we don't get false test failures.

bosonie replied:
Ok, I will revert the changes and I will put a huge number for the tolerances in my case.

greschd replied:
> a huge number

I mean, only needs to be such that the algorithm can't reasonably be outside of this bound.

I'd say, take the highest deviation you find in e.g. 10 runs, multiply it by ~10, that should be big enough. And still give us some indication that the algorithm went to the right place.

greschd commented on aiida_optimize/engines/_particle_swarm.py:
```diff
@@ -0,0 +1,268 @@
+
+    @staticmethod
+    def update_vel(omega, v, c1, c2, x, pi, pg):  # pylint: disable=too-many-arguments
+        return omega * v + c1 * uniform(0, 1) * (pi - x) + c2 * uniform(0, 1) * (pg - x)
```
I wonder if we need to add some logic to ensure we get "good" randomness behavior here. I'm more familiar with Monte Carlo type calculations, but there one would typically seed the RNG with a known number, to get reproducible results.

Here, the trouble is that each step can happen from a new / random Python interpreter. This presents a challenge because we can't simply rely on the RNG remaining the same. So even if we don't care about reproducibility, it could mean that we don't get "good" randomness.
I suspect a solution to this would include either persisting the RNG state between steps (maybe `numpy.random` can help with this?), or re-seeding and then re-iterating through the previous values (keeping track of how many random numbers we used).

bosonie replied:
After discussion with @vdikan, we implemented to pass the state of the RNG among steps, using `get_state` and `set_state` of `numpy.random`. Let us know what you think.

greschd replied:
Sweet, that looks good to me! :+1:

greschd commented on aiida_optimize/engines/_particle_swarm.py:
```diff
@@ -0,0 +1,268 @@
+# -*- coding: utf-8 -*-
+
+# The additional license terms given in ADDITIONAL_TERMS.txt apply to this
```
This is also part of the scipy license header that can be removed.

bosonie committed:
Improve randomness

1) add random velocoties for the initial sterp
2) pass the state of the random generator from one iteration to the other

bosonie committed:
Remove optional xtol and ftol in tests

We leave it and set big number for the ParticleSwarm

bosonie committed:
Fix linter

bosonie commented:
@greschd should be ready!

greschd commented:
Good stuff, thanks @bosonie!

greschd commented:
Released as v0.5.0: https://pypi.org/project/aiida-optimize/0.5.0/. Please let me know in case you find any issues in the `aiida-siesta` testing!
